broker.go

```diff
@@ -0,0 +1,706 @@
+type Consumer struct {
+	broker *Broker
+	conn   *connection
+	conf   ConsumerConf
+	offset int64
+	msgbuf []*proto.Message
+}
+
+// Consumer creates new consumer instance, bound to broker.
+func (b *Broker) Consumer(conf ConsumerConf) (consumer *Consumer, err error) {
+	conn, err := b.leaderConnection(conf.Topic, conf.Partition)
+	if err != nil {
+		return nil, err
+	}
+	if conf.Log == nil {
+		conf.Log = b.conf.Log
+	}
+	offset := conf.StartOffset
+	if conf.StartOffset < 0 {
+		switch conf.StartOffset {
+		case StartOffsetNewest:
+			off, err := b.OffsetLatest(conf.Topic, conf.Partition)
+			if err != nil {
+				return nil, err
+			}
+			// latest offset that next produced message will get, but consumer
+			// offset should be set to the last returned message offset
+			offset = off - 1
+		case StartOffsetOldest:
+			off, err := b.OffsetEarliest(conf.Topic, conf.Partition)
+			if err != nil {
+				return nil, err
+			}
+			offset = off
+			// this is stange kafka thing: asking about message with offset one
+			// higher than the last existing message is fine, except when there
+			// are no messages in the log -- in such case you should always ask
+			// about offset 0.
+			// Set offset -1 to make fetch request 0 offset message when there
+			// are none.
+			if off == 0 {
+				offset = -1
+			}
+		default:
+			return nil, fmt.Errorf("invalid start offset: %d", conf.StartOffset)
+		}
+	}
+	consumer = &Consumer{
+		broker: b,
+		conn:   conn,
+		conf:   conf,
+		msgbuf: make([]*proto.Message, 0),
+		offset: offset,
+	}
+	return consumer, nil
+}
+
+// Conf returns consumer configuration.
+func (c *Consumer) Conf() ConsumerConf {
+	return c.conf
+}
+
+// Fetch is returning single message from consumed partition. Consumer can
+// retry fetching messages even if responses return no new data. Retry
+// behaviour can be configured through RetryLimit and RetryWait consumer
+// parameters.
+//
+// Fetch is not thread safe.
+//
+// Fetch can retry sending request on common errors. This behaviour can be
+// configured with RetryErrLimit and RetryErrWait consumer configuration
+// attributes.
+func (c *Consumer) Fetch() (*proto.Message, error) {
+	var retry int
+
+	for len(c.msgbuf) == 0 {
+		messages, err := c.fetch()
+		if err != nil {
+			return nil, err
+		}
+
+		// check first messages if any of them has index lower than requested,
+		// this can happen because of kafka optimizations
+		toSkip := 0
+		for toSkip < len(messages) {
+			if messages[toSkip].Offset >= c.offset {
+				break
+			}
+			toSkip++
+		}
+
+		// messages crc is checked by underlying connection so no need to check
+		// those again
+
+		if len(messages) == toSkip {
+			c.conf.Log.Printf("none of %d fetched messages is valid", toSkip)
+		}
+		// ignore all messages that are of index lower than requested
+		c.msgbuf = messages[toSkip:]
+
+		if len(c.msgbuf) == 0 {
+			time.Sleep(time.Duration(math.Log(float64(retry+2))) * c.conf.RetryWait)
```

Comment:
Does Kafka documentation explain how to use long polling? This would be better here than logarithmic retries.